include/grpc/grpc_security.h

```diff
@@ -834,30 +838,73 @@ typedef void (*grpc_tls_on_credential_reload_done_cb)(
+    field with labels. Here are their explanations
+    -------------------------label explanation----------------------------------
+    [USER-MANAGED]: fields that are managed directly by the end users
+    [SYSTEM-PROVIDED]: fields that needed to retrieve implementations in C core
+    and would be provided by the gRPC stack
+    [WRAP-LANG]: fields that needed to retrieve the particular implementation in
+    wrap language
+    [ASYNC_ONLY]: fields that only used when in asynchronous mode. If a field
+    doesn't have this label, we can assume it should be applied to both
+    synchronous and asynchronous mode.
+    Note that if a field is marked both as [USER-MANAGED] and [SYSTEM-PROVIDED],
+    it indicates that the field is initiated by the gRPC stack, but users
+    need to properly manage them to correctly perform credential reloading.
+    ----------------------label explanation ends--------------------------------
+    - cb[USER-MANAGED][ASYNC_ONLY]: a callback-function provided for users to
+      indicate an asynchronous reloading operation is complete. After calling
+      this function, gRPC stack would be notified and continue rest of the
+      authentication.
+    - cb_user_data[SYSTEM-PROVIDED][ASYNC_ONLY]: a pointer used to retrieve the
+      corresponding security connector implementation. This is needed because
+      we need to use the |peer_checked| closure in security connector after
+      users call |cb|.
+    - key_materials_config[USER-MANAGED][SYSTEM-PROVIDED]: an in/output
+      parameter containing currently used/newly reloaded credentials. Users can
+      assume it always points to a non-null object.
+      If users set |status| to |GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_UNCHANGED|,
+      key_materials_config should not be modified.
+      If set to |GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_NEW|, key_materials_config
+      object can be updated by the new credentials.
+      Note that, key_materials_config might be shared by multiple connections,
+      because each key_materials_config is attached to a security_connector, and
+          1. on client side, if retry mechanism happens, one security_connector
+          can map to multiple connections
+          2. on server side, all connections share the same security_connector
+      But the reload logic is guaranteed to be invoked in every connection.
+    - status[USER-MANAGED]: a per-connection parameter to indicate the finish
+      status of a schedule/cancel request.
+      If set to |GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_UNCHANGED|, gRPC stack
+      would take current |grpc_tls_key_materials_config| as credentials;
+      If set to |GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_NEW|, gRPC stack would
+      perform credential reloading, using newly loaded |key_materials_config|;
+      If set to |GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_FAIL|, users also need to
+      set |error_details|.
+    - err_details[USER-MANAGED][SYSTEM-PROVIDED]: a per-connection
```

Comment:
`SYSTEM-PROVIDED` ? I thought this is set by the callback to indicate the error happened. This needs to be clarified because it is confusing. Also `grpc_tls_err_details` struct is defined in a different file `grpc_tls_credentials_options.h` but if the callback is going to set the error string in the object they also need access to that definition, right? My point is `grpc_tls_err_details` is not just an opaque struct that callers don't need to know about.

Reply:
A suggestion: both status and err_details are used together in `grpc_tls_credential_reload_arg` and `grpc_tls_server_authorization_check_arg` . Why not make status part of `grpc_tls_err_details` (and rename it to `grpc_tls_status_details`) so have one less field in both these complex structs?

Reply:
I've thought this before but it turned out that the `grpc_tls_credential_reload_arg`'s status is using  a type of `grpc_ssl_certificate_config_reload_status` but `grpc_tls_server_authorization_check_arg` is using a type of `grpc_status_code`.

Reply:
> I thought this is set by the callback to indicate the error happened. This needs to be clarified because it is confusing. Also grpc_tls_err_details struct is defined in a different file grpc_tls_credentials_options.h but if the callback is going to set the error string in the object they also need access to that definition, right?

Yes, I marked it both as  [USER-MANAGED] and [SYSTEM-PROVIDED], because from a user's point of view, this field should be managed properly to indicate a failure;  from a maintainer's pointer of view, this field should be properly set in gRPC stack to give it for users to use. I was mixing the semantics here. Probably using a label is not a good idea...

Reply:
I thought labels was a step in the right direction but might need some tweaking so that might be worth trying?

Reply:
I pushed another version without the label. Maybe we can compare the two versions and find a better one?